### PR TITLE
BACK 1169 - Testnet handled by Lama

### PR DESCRIPTION
**Pull request by mbasire-ledger**
## What is this about?

Testnet has to be handled by Lama. It will be the first fork to be handled.

## Cute picture of lama
lama

### CODE REVIEW

File: coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Interpreter.scala [hunk 1]
```diff
@@ -160,7 +160,7 @@ class DbInterpreter(
               OperationNotification(
                 accountId = accountId,
                 coinFamily = CoinFamily.Bitcoin,
-                coin = Coin.Btc,
+                coin = Coin.fromKey(request.coinId).get,
```
Comment (nukky94) [inline coins/bitcoin/interpreter/src/main/scala/co/ledger/lama/bitcoin/interpreter/Interpreter.scala:163]: better to validate and parse correctly the request

**Review by nukky94 — CHANGES_REQUESTED** [review]

File: account-manager/src/main/resources/db/migration/V2__coin_type.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TYPE coin ADD VALUE 'btc_testnet' ;
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply account-manager/src/main/resources/db/migration/V2__coin_type.sql]: ok

File: coins/bitcoin/api/src/it/resources/application.conf [hunk 3]
```diff
@@ -5,3 +5,56 @@ server = {
   port = 8080
   port = ${?BITCOIN_API_PORT}
 }
+
+events-config {
+
+  lama-events-exchange-name = "lama.notifications"
+  lama-events-exchange-name = ${?LAMA_NOTIFICATIONS_EXCHANGE_NAME}
+
+    coins {
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/it/resources/application.conf]: ok

File: coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala [hunk 4]
```diff
@@ -234,3 +259,47 @@ class AccountControllerIT extends AnyFlatSpecLike with Matchers {
...
+            accountId,
+            coinConf.coinFamily,
+            coinConf.coin
+          )
+        )
+      }
+
+  }
+
+  def waitBalancedUpdated(
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala]: ok

File: coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala [hunk 5]
```diff
@@ -4,103 +4,124 @@ import java.time.Instant
...
+              accountRegistered.accountId,
+              conf.eventsConfig.coins(account.registerRequest.coin),
+              notificationService
+            )
+
             accountInfoAfterRegister <- client.expect[AccountInfo](
               getAccountRequest(accountRegistered.accountId)
             )
 
+            balancedNotification <- AccountNotifications.waitBalancedUpdated(notifications)
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala]: ok

File: coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala [hunk 6]
```diff
@@ -182,6 +203,10 @@ class AccountControllerIT extends AnyFlatSpecLike with Matchers {
                 .map(_.status) should (contain(Registered) or contain(Published))
             }
 
+            it should "emit a balanced notification" in {
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala]: ok

File: coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala [hunk 7]
```diff
@@ -182,6 +203,10 @@ class AccountControllerIT extends AnyFlatSpecLike with Matchers {
...
+              balancedNotification.accountId shouldBe accountRegistered.accountId
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/AccountControllerIT.scala]: ok

File: coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/ConfigSpec.scala [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  case class ServerConfig(
+                           host: String,
+                           port: Int
+                         )
+
+  case class EventsConfig(
+                           lamaEventsExchangeName: ExchangeName,
+                           rabbit: Fs2RabbitConfig,
+                           coins: Map[Coin, CoinConfig]
+                         )
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/it/scala/co/ledger/lama/bitcoin/api/ConfigSpec.scala]: ok

File: coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/utils/ProtobufUtils.scala [hunk 9]
```diff
@@ -32,9 +32,21 @@ object ProtobufUtils {
   def toCoin(c: Coin): co.ledger.lama.manager.protobuf.Coin =
     c match {
       case Coin.Btc => co.ledger.lama.manager.protobuf.Coin.btc
+      case Coin.BtcTestnet => co.ledger.lama.manager.protobuf.Coin.btc_testnet
       case _        => co.ledger.lama.manager.protobuf.Coin.Unrecognized(-1)
     }
 
+  // do we need to handle Unrecognized ?
```
[the remark this answers is not in the document]
Comment (mbasire-ledger) [reply coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/utils/ProtobufUtils.scala]: ok

**Comment by mbasire-ledger** [comment]
Lama bitcoin api is able to handle register request for testnet. The testnet  account will be synchronized.

**Review by nukky94 — APPROVED** [review]
Good work! LGTM

**Comment by nukky94** [comment]
TODO:
- Add your gpg key on Github > Settings > SSH and GPG keys
- Merge master and [Sign your commit](https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/signing-commits)

**Pull request closed without merging**